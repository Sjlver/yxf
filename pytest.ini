[pytest]
addopts = --doctest-modules --ignore=Lib