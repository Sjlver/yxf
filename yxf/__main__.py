"""
yxf: Convert from XLSForm to YAML and back.

To convert an XLSForm to a YAML file: `python -m yxf form.xlsx`.

By default, the result will be called `form.yaml`, in other words, the same name
as the input file with the extension changed to `.yaml`. You can specify a
different output file name using the `--output othername.yaml` option.

To convert a YAML file to an XLSForm: `python -m yxf form.yaml`.
"""

import argparse
import collections
import logging
import pathlib

import openpyxl
import openpyxl.styles
import openpyxl.utils
import strictyaml

from . import xlsform

log = logging.getLogger("yxf.__main__")


def _convert_sheet(sheet):
    headers = xlsform.headers(sheet)
    result = []
    for row in xlsform.content_rows(sheet, values_only=True):
        values = xlsform.truncate_row(row)
        row_dict = collections.OrderedDict()
        for h, v in zip(headers, values):
            if v is None:
                continue
            if h is None:
                raise ValueError(f"Cell with no column header: {v}")
            row_dict[h] = v
        if row_dict:
            result.append(row_dict)
    return result


def _convert_to_sheet(sheet, rows, keys):
    key_set = set(keys)

    for i, key in enumerate(keys):
        sheet.cell(row=1, column=i + 1, value=key)

    next_row = 2
    for row in rows:
        if row.get("type") == "begin_group":
            next_row += 1

        if not all(k in key_set for k in row.keys()):
            missing_key = next(k for k in row.keys() if k not in key_set)
            raise ValueError(
                f'Invalid key "{missing_key}" in row "{row.get("name", "(unnamed)")}". Add it to yxf.headers.{sheet.title} in the YAML file.'
            )

        for i, key in enumerate(keys):
            if key in row:
                sheet.cell(row=next_row, column=i + 1, value=row[key])

        next_row += 1

    return sheet


def _check_existing_output(filename, force):
    if filename.exists() and not force:
        raise ValueError(f"File already exists (use --force to override): {filename}")


def _ensure_yxf_comment(form, name):
    desired_comment = (
        f"Converted by yxf, from {name}. Edit the YAML file instead of the Excel file."
    )
    first_line = form["survey"][0]
    if "#" not in first_line or not first_line["#"].startswith("Converted by yxf,"):
        form["survey"].insert(0, {"#": desired_comment})
    else:
        form["survey"][0]["#"] = desired_comment


def xlsform_to_yaml(filename: pathlib.Path, target: pathlib.Path):
    """Convert XLSForm file `filename` to YAML file `target`."""

    log.info("xlsform_to_yaml: %s -> %s", filename, target)

    wb = openpyxl.load_workbook(filename, read_only=True)
    result = collections.OrderedDict()
    headers = collections.OrderedDict()
    for sheet_name in ["survey", "choices", "settings"]:
        if sheet_name in wb:
            result[sheet_name] = _convert_sheet(wb[sheet_name])
            headers[sheet_name] = xlsform.headers(wb[sheet_name])
            if headers[sheet_name] and headers[sheet_name][0] != "#":
                if "#" in headers[sheet_name]:
                    raise ValueError(
                        f"The comment column must come first in sheet {sheet_name}."
                    )
                headers[sheet_name].insert(0, "#")

    if "survey" not in result:
        raise ValueError('An XLSForm must have a "survey" sheet.')

    _ensure_yxf_comment(result, filename.name)
    result["yxf"] = {"headers": headers}

    with open(target, "w", encoding="utf-8") as f:
        f.write(strictyaml.as_document(result).as_yaml())


def xlsform_to_markdown(filename: pathlib.Path, target: pathlib.Path):
    """Convert XLSForm file `filename` to Markdown file `target`."""

    log.info("xlsform_to_markdown: %s -> %s", filename, target)

def yaml_to_xlsform(filename: pathlib.Path, target: pathlib.Path):
    """Convert YAML file `filename` to XLSForm file `target`."""

    log.info("yaml_to_xlsform: %s -> %s", filename, target)

    with open(filename, encoding="utf-8") as f:
        form = strictyaml.load(f.read()).data

    if "yxf" not in form:
        raise ValueError('YAML file must have a "yxf" entry.')
    if "survey" not in form:
        raise ValueError('YAML file must have a "survey" entry.')
    _ensure_yxf_comment(form, filename.name)

    wb = openpyxl.Workbook()
    for sheet_name in form:
        if sheet_name == "yxf":
            continue
        _convert_to_sheet(
            wb.create_sheet(sheet_name),
            form[sheet_name],
            form["yxf"]["headers"][sheet_name],
        )
    wb.remove(wb.active)
    xlsform.make_pretty(wb)
    wb.save(target)

def markdown_to_xlsform(filename: pathlib.Path, target: pathlib.Path):
    """Convert Markdown file `filename` to XLSForm file `target`."""

    log.info("markdown_to_xlsform: %s -> %s", filename, target)

def main():
    logging.basicConfig(level=logging.DEBUG)

    parser = argparse.ArgumentParser(
        description="Convert from XLSForm to YAML and back"
    )
    parser.add_argument("file", type=pathlib.Path, help="a file to be converted")
    parser.add_argument(
        "--markdown",
        action="store_true",
        help="use Markdown instead of YAML",
    )
    parser.add_argument(
        "-o",
        "--output",
        type=pathlib.Path,
        help="output file name (default: same as input, with extension changed)",
    )
    parser.add_argument(
        "-f",
        "--force",
        action="store_true",
        help="allow overwriting existing output files",
    )
    args = parser.parse_args()

    if args.file.suffix == ".xlsx":
        if args.markdown:
            args.output = args.output or args.file.with_suffix(".md")
            _check_existing_output(args.output, args.force)
            xlsform_to_markdown(args.file, args.output)
        else:
            args.output = args.output or args.file.with_suffix(".yaml")
            _check_existing_output(args.output, args.force)
            xlsform_to_yaml(args.file, args.output)
    elif args.file.suffix == ".yaml":
        args.output = args.output or args.file.with_suffix(".xlsx")
        _check_existing_output(args.output, args.force)
        yaml_to_xlsform(args.file, args.output)
    elif args.file.suffix == ".md":
        args.output = args.output or args.file.with_suffix(".xlsx")
        _check_existing_output(args.output, args.force)
        markdown_to_xlsform(args.file, args.output)
    else:
        raise ValueError(f"Unrecognized file extension: {args.file}")


if __name__ == "__main__":
    main()
