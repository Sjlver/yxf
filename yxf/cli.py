"""Command-line interface for yxf.

This module handles file I/O and CLI argument parsing, delegating the actual
conversion work to the library modules (excel, yaml, markdown).
"""

import argparse
import logging
import pathlib

from . import excel, markdown, yaml

log = logging.getLogger(__name__)


def _check_existing_output(filename: pathlib.Path, force: bool) -> None:
    """Check if output file exists and raise error if it does (unless force=True).

    Args:
        filename: Path to output file
        force: If True, allow overwriting existing files

    Raises:
        ValueError: If file exists and force is False
    """
    if filename.exists() and not force:
        raise ValueError(f"File already exists (use --force to override): {filename}")


def xlsform_to_yaml(filename: pathlib.Path, target: pathlib.Path):
    """Convert XLSForm file to YAML file.

    Args:
        filename: Path to input Excel file
        target: Path to output YAML file
    """
    log.info("xlsform_to_yaml: %s -> %s", filename, target)

    with open(filename, "rb") as f:
        form = excel.read_xlsform(f)

    excel.ensure_yxf_comment(form, filename.name, "YAML")
    yaml_content = yaml.write_yaml(form)

    with open(target, "w", encoding="utf-8") as f:
        f.write(yaml_content)


def xlsform_to_markdown(filename: pathlib.Path, target: pathlib.Path):
    """Convert XLSForm file to Markdown file.

    Args:
        filename: Path to input Excel file
        target: Path to output Markdown file
    """
    log.info("xlsform_to_markdown: %s -> %s", filename, target)

    with open(filename, "rb") as f:
        form = excel.read_xlsform(f)

    excel.ensure_yxf_comment(form, filename.name, "Markdown")
    md_content = markdown.write_markdown(form, filename.name)

    with open(target, "w", encoding="utf-8") as f:
        f.write(md_content)


def yaml_to_xlsform(filename: pathlib.Path, target: pathlib.Path):
    """Convert YAML file to XLSForm file.

    Args:
        filename: Path to input YAML file
        target: Path to output Excel file
    """
    log.info("yaml_to_xlsform: %s -> %s", filename, target)

    with open(filename, encoding="utf-8") as f:
        form = yaml.read_yaml(f.read())

    excel.ensure_yxf_comment(form, filename.name, "YAML")

    with open(target, "wb") as f:
        excel.write_xlsform(form, f)


def markdown_to_xlsform(filename: pathlib.Path, target: pathlib.Path):
    """Convert Markdown file to XLSForm file.

    Args:
        filename: Path to input Markdown file
        target: Path to output Excel file
    """
    log.info("markdown_to_xlsform: %s -> %s", filename, target)

    with open(filename, encoding="utf-8") as f:
        form = markdown.read_markdown(f.read(), filename.name)

    excel.ensure_yxf_comment(form, filename.name, "Markdown")

    with open(target, "wb") as f:
        excel.write_xlsform(form, f)


def main():
    """yxf: Convert from XLSForm to YAML and back."""

    logging.basicConfig(level=logging.DEBUG)
    logging.getLogger("markdown_it").setLevel(logging.INFO)

    parser = argparse.ArgumentParser(
        description="Convert from XLSForm to YAML and back"
    )
    parser.add_argument("file", type=pathlib.Path, help="a file to be converted")
    parser.add_argument(
        "--markdown",
        action="store_true",
        help="use Markdown instead of YAML",
    )
    parser.add_argument(
        "-o",
        "--output",
        type=pathlib.Path,
        help="output file name (default: same as input, with extension changed)",
    )
    parser.add_argument(
        "-f",
        "--force",
        action="store_true",
        help="allow overwriting existing output files",
    )
    args = parser.parse_args()

    if args.file.suffix == ".xlsx":
        if args.markdown or (args.output and args.output.suffix == ".md"):
            args.output = args.output or args.file.with_suffix(".md")
            _check_existing_output(args.output, args.force)
            xlsform_to_markdown(args.file, args.output)
        else:
            args.output = args.output or args.file.with_suffix(".yaml")
            _check_existing_output(args.output, args.force)
            xlsform_to_yaml(args.file, args.output)
    elif args.file.suffix == ".yaml":
        args.output = args.output or args.file.with_suffix(".xlsx")
        _check_existing_output(args.output, args.force)
        yaml_to_xlsform(args.file, args.output)
    elif args.file.suffix == ".md":
        args.output = args.output or args.file.with_suffix(".xlsx")
        _check_existing_output(args.output, args.force)
        markdown_to_xlsform(args.file, args.output)
    else:
        raise ValueError(f"Unrecognized file extension: {args.file}")
