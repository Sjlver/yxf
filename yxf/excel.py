"""Functions for reading and writing XLSForm Excel files.

This module provides the core library functions for converting between XLSForm
Excel files and Python dictionaries. Functions accept file-like objects (BinaryIO)
rather than filenames for maximum flexibility.
"""

import collections
import logging
from typing import BinaryIO

import openpyxl

from . import xlsform

log = logging.getLogger(__name__)


def row_to_dict(headers, values):
    """Convert a row of values to an ordered dictionary using headers as keys.

    Args:
        headers: List of column headers
        values: List of cell values

    Returns:
        OrderedDict with non-empty values mapped to their headers

    Raises:
        ValueError: If a non-empty value has no corresponding header

    >>> row_to_dict(["name", "type", "label"], ["q1", "text", "Question 1"])
    OrderedDict({'name': 'q1', 'type': 'text', 'label': 'Question 1'})
    >>> row_to_dict(["name", "type"], ["q1", ""])
    OrderedDict({'name': 'q1'})
    >>> row_to_dict(["name", "type"], ["q1", None])
    OrderedDict({'name': 'q1'})
    """
    row_dict = collections.OrderedDict()
    for h, v in zip(headers, values):
        if v is None or v == "":
            continue
        if h is None:
            raise ValueError(f"Cell with no column header: {v}")
        row_dict[h] = v
    return row_dict


def _convert_sheet(sheet):
    """Convert an Excel sheet to a list of dictionaries.

    Args:
        sheet: openpyxl Worksheet object

    Returns:
        List of OrderedDicts, one per row (excluding header)
    """
    headers = xlsform.headers(sheet)
    result = []
    for row in xlsform.content_rows(sheet, values_only=True):
        values = xlsform.truncate_row(row)
        values = [xlsform.stringify_value(v) for v in values]
        row_dict = row_to_dict(headers, values)
        if row_dict:
            result.append(row_dict)
    return result


def _convert_to_sheet(sheet, rows, keys):
    """Convert a list of dictionaries to an Excel sheet.

    Args:
        sheet: openpyxl Worksheet object to write to
        rows: List of OrderedDicts representing rows
        keys: List of column headers

    Returns:
        The modified sheet

    Raises:
        ValueError: If a row contains a key not in the headers list
    """
    key_set = set(keys)

    for i, key in enumerate(keys):
        sheet.cell(row=1, column=i + 1, value=key)

    next_row = 2
    previous_list_name = rows[0].get("list_name") if rows else None
    for row in rows:
        if row.get("type") == "begin_group":
            next_row += 1

        if row.get("list_name") != previous_list_name:
            previous_list_name = row.get("list_name")
            next_row += 1

        if not all(k in key_set for k in row.keys()):
            missing_key = next(k for k in row.keys() if k not in key_set)
            raise ValueError(
                f'Invalid key "{missing_key}" in row "{row.get("name", "(unnamed)")}". '
                f"Add it to yxf.headers.{sheet.title} in the YAML file."
            )

        for i, key in enumerate(keys):
            if key in row:
                sheet.cell(row=next_row, column=i + 1, value=row[key])

        next_row += 1

    return sheet


def validate_sheet_name(sheet_name, source_name, line):
    """Validate that a sheet name is one of the allowed XLSForm sheets.

    Args:
        sheet_name: Name to validate
        source_name: Name of source file (for error messages)
        line: Line number (for error messages)

    Raises:
        ValueError: If sheet name is not valid

    >>> validate_sheet_name("survey", "test.yaml", 1)
    >>> validate_sheet_name("choices", "test.yaml", 1)
    >>> validate_sheet_name("settings", "test.yaml", 1)
    >>> validate_sheet_name("invalid", "test.yaml", 1)
    Traceback (most recent call last):
        ...
    ValueError: test.yaml:1: Invalid sheet name (must be survey, choices, or settings): invalid
    """
    if sheet_name not in ["survey", "choices", "settings"]:
        raise ValueError(
            f"{source_name}:{line}: Invalid sheet name (must be survey, choices, or settings): {sheet_name}"
        )


def ensure_yxf_comment(form, name, file_format):
    """Ensure the form has a yxf conversion comment in the first row.

    Args:
        form: Form dictionary to modify
        name: Name of source file
        file_format: Format name (e.g., "YAML", "Markdown")
    """
    desired_comment = f"Converted by yxf, from {name}. Edit the {file_format} file instead of the Excel file."

    first_line = form["survey"][0]
    if "#" not in first_line or not first_line["#"].startswith("Converted by yxf,"):
        form["survey"].insert(0, {"#": desired_comment})
    else:
        form["survey"][0]["#"] = desired_comment

    if "#" not in form["yxf"]["headers"]["survey"]:
        form["yxf"]["headers"]["survey"].insert(0, "#")


def read_xlsform(file_obj: BinaryIO) -> dict:
    """Read an XLSForm from a file-like object.

    Args:
        file_obj: Binary file-like object containing Excel data

    Returns:
        Dictionary with sheet names as keys and lists of row dicts as values.
        Also includes a "yxf" key with metadata including headers.

    Raises:
        ValueError: If the workbook doesn't have a "survey" sheet
    """
    wb = openpyxl.load_workbook(file_obj, read_only=True)
    result = collections.OrderedDict()
    headers = collections.OrderedDict()

    for sheet_name in ["survey", "choices", "settings"]:
        if sheet_name in wb:
            result[sheet_name] = _convert_sheet(wb[sheet_name])
            headers[sheet_name] = xlsform.headers(wb[sheet_name])
            if headers[sheet_name] and headers[sheet_name][0] != "#":
                if "#" in headers[sheet_name]:
                    raise ValueError(
                        f"The comment column must come first in sheet {sheet_name}."
                    )

    if "survey" not in result:
        raise ValueError('An XLSForm must have a "survey" sheet.')

    result["yxf"] = {"headers": headers}
    return result


def write_xlsform(form: dict, file_obj: BinaryIO) -> None:
    """Write a form dictionary to an XLSForm Excel file.

    Args:
        form: Dictionary with sheet names as keys and lists of row dicts as values.
              Must include a "yxf" key with headers metadata.
        file_obj: Binary file-like object to write Excel data to

    Raises:
        ValueError: If a row contains a key not in the sheet's headers
    """
    wb = openpyxl.Workbook()
    for sheet_name in form:
        if sheet_name == "yxf":
            continue
        _convert_to_sheet(
            wb.create_sheet(sheet_name),
            form[sheet_name],
            form["yxf"]["headers"][sheet_name],
        )
    if wb.active is not None:
        wb.remove(wb.active)
    xlsform.make_pretty(wb)
    wb.save(file_obj)
