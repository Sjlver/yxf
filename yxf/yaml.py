"""Functions for reading and writing YAML format XLSForms.

This module provides functions for converting between YAML strings and
XLSForm dictionaries. Functions work with strings rather than file objects
for maximum flexibility.
"""

import strictyaml


def read_yaml(content: str) -> dict:
    """Parse a YAML string into a form dictionary.

    Args:
        content: YAML string content

    Returns:
        Dictionary with sheet names as keys and lists of row dicts as values.
        Includes a "yxf" key with metadata.

    Raises:
        ValueError: If YAML is missing required keys
        strictyaml.YAMLError: If YAML is malformed
    """
    form = strictyaml.load(content).data

    if "yxf" not in form:
        raise ValueError('YAML file must have a "yxf" entry.')
    if "survey" not in form:
        raise ValueError('YAML file must have a "survey" entry.')

    return form


def write_yaml(form: dict) -> str:
    """Convert a form dictionary to a YAML string.

    Args:
        form: Dictionary with sheet names as keys and lists of row dicts as values.
              Must include a "yxf" key with headers metadata.

    Returns:
        YAML string representation of the form
    """
    return strictyaml.as_document(form).as_yaml()
