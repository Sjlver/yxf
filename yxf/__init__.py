from .__main__ import (
    xlsform_bytes_to_yaml_string,
    xlsform_bytes_to_markdown_string,
    yaml_string_to_xlsform_bytes,
    markdown_string_to_xlsform_bytes,
)

__all__ = [
    'xlsform_bytes_to_yaml_string',
    'xlsform_bytes_to_markdown_string',
    'yaml_string_to_xlsform_bytes',
    'markdown_string_to_xlsform_bytes',
]
