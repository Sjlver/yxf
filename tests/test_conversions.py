"""Snapshot tests for Excel ↔ YAML ↔ Markdown conversions.

These tests use syrupy for snapshot testing to ensure conversion outputs
remain stable. Only YAML and Markdown outputs are snapshotted (not Excel
files, as openpyxl output is not byte-stable).
"""

import io
import pathlib

import pytest

from yxf import (
    read_xlsform,
    write_xlsform,
    read_yaml,
    write_yaml,
    read_markdown,
    write_markdown,
)


# Test data directory
TESTDATA_DIR = pathlib.Path(__file__).parent / "testdata"


@pytest.fixture(
    params=[
        "favorite-color.xlsx",
        "simple-repeat.xlsx",
        "cascading-select.xlsx",
        "xlsform-dot-org-template.xlsx",
    ]
)
def excel_file(request):
    """Fixture providing all test Excel files."""
    return TESTDATA_DIR / request.param


def test_excel_to_yaml_snapshot(excel_file, snapshot):
    """Test Excel → YAML conversion produces stable output."""
    with open(excel_file, "rb") as f:
        form = read_xlsform(f)

    # Skip xlsform.org template - it has empty sheets that strictyaml can't serialize
    if "xlsform-dot-org-template" in excel_file.name:
        pytest.skip(
            "xlsform.org template has empty sheets that strictyaml cannot serialize"
        )

    yaml_output = write_yaml(form)
    assert yaml_output == snapshot


def test_excel_to_markdown_snapshot(excel_file, snapshot):
    """Test Excel → Markdown conversion produces stable output."""
    with open(excel_file, "rb") as f:
        form = read_xlsform(f)
    markdown_output = write_markdown(form, excel_file.name)
    assert markdown_output == snapshot


def test_yaml_roundtrip_stability(excel_file):
    """Test Excel → YAML → Excel → YAML produces identical YAML.

    This verifies that converting through Excel doesn't lose or alter data.
    """
    # Skip xlsform.org template - it has empty sheets that strictyaml can't serialize
    if "xlsform-dot-org-template" in excel_file.name:
        pytest.skip(
            "xlsform.org template has empty sheets that strictyaml cannot serialize"
        )

    # Excel → YAML
    with open(excel_file, "rb") as f:
        form1 = read_xlsform(f)
    yaml1 = write_yaml(form1)

    # YAML → Excel (in-memory) → YAML
    form2 = read_yaml(yaml1)
    excel_bytes = io.BytesIO()
    write_xlsform(form2, excel_bytes)
    excel_bytes.seek(0)
    form3 = read_xlsform(excel_bytes)
    yaml2 = write_yaml(form3)

    assert yaml1 == yaml2, "YAML should be identical after round-trip through Excel"


def test_markdown_roundtrip_via_yaml(excel_file):
    """Test Excel → Markdown → Excel → YAML produces same result as Excel → YAML.

    This verifies that Markdown conversion preserves all data needed for YAML.
    Note: We compare via YAML since that's our canonical format.
    """
    # Skip xlsform.org template - it has empty sheets that strictyaml can't serialize
    if "xlsform-dot-org-template" in excel_file.name:
        pytest.skip(
            "xlsform.org template has empty sheets that strictyaml cannot serialize"
        )

    # Excel → YAML (baseline)
    with open(excel_file, "rb") as f:
        form1 = read_xlsform(f)
    yaml_baseline = write_yaml(form1)

    # Excel → Markdown → Excel → YAML
    markdown = write_markdown(form1, excel_file.name)
    form2 = read_markdown(markdown, excel_file.name)
    excel_bytes = io.BytesIO()
    write_xlsform(form2, excel_bytes)
    excel_bytes.seek(0)
    form3 = read_xlsform(excel_bytes)
    yaml_after_markdown = write_yaml(form3)

    # Compare the YAMLs (they should be very similar, though comments may differ)
    assert yaml_baseline == yaml_after_markdown

    # We check that the structure is preserved
    assert "survey" in yaml_after_markdown
    assert "yxf" in yaml_after_markdown
