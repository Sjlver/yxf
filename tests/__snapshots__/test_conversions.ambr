# serializer version: 1
# name: test_excel_to_markdown_snapshot[cascading-select.xlsx]
  '''
  ## survey
  
  | type                | name            | label           | choice_filter             |
  | ------------------- | --------------- | --------------- | ------------------------- |
  | select_one states   | selected_state  | Select a state  |                           |
  | select_one counties | selected_county | Select a county | state=${selected_state}   |
  | select_one cities   | selected_city   | Select a city   | county=${selected_county} |
  
  ## choices
  
  | list_name | name        | label       | state      | county  |
  | --------- | ----------- | ----------- | ---------- | ------- |
  | states    | texas       | Texas       |            |         |
  | states    | washington  | Washington  |            |         |
  | counties  | king        | King        | washington |         |
  | counties  | pierce      | Pierce      | washington |         |
  | counties  | brewster    | Brewster    | texas      |         |
  | counties  | cameron     | Cameron     | texas      |         |
  | cities    | brownsville | Brownsville | texas      | cameron |
  | cities    | harlingen   | Harlingen   | texas      | cameron |
  | cities    | seattle     | Seattle     | washington | king    |
  | cities    | redmond     | Redmond     | washington | king    |
  | cities    | tacoma      | Tacoma      | washington | pierce  |
  | cities    | puyallup    | Puyallup    | washington | pierce  |
  
  ## settings
  
  | form_title               | form_id          |
  | ------------------------ | ---------------- |
  | Cascading select example | cascading_select |
  
  '''
# ---
# name: test_excel_to_markdown_snapshot[favorite-color.xlsx]
  '''
  ## survey
  
  | type              | name                 | label::English (en)                | label::French (fr)                        | required | relevant                         | parameters                                                               | appearance |
  | ----------------- | -------------------- | ---------------------------------- | ----------------------------------------- | -------- | -------------------------------- | ------------------------------------------------------------------------ | ---------- |
  | text              | first_name           | What is your first name?           | Quel est votre prénom?                    | true()   |                                  |                                                                          |            |
  | integer           | age                  | What is your age, ${first_name}?   | Quel âge as-tu, ${first_name}?            | true()   |                                  |                                                                          |            |
  | select_one colors | favorite_color       | What is your favorite color?       | Quelle est votre couleur préférée?        |          |                                  |                                                                          |            |
  | text              | favorite_color_other | Please specify your favorite color | Veuillez spécifier votre couleur préférée |          | selected(${favorite_color}, 'o') |                                                                          |            |
  | geopoint          | location             | Record your location               | Enregistrez votre position                | true()   | ${age} >= 18                     |                                                                          | hide-input |
  | audit             | audit                |                                    |                                           |          |                                  | location-priority=balanced location-min-interval=60 location-max-age=120 |            |
  
  ## choices
  
  | list_name | name | label::English (en) | label::French (fr) |
  | --------- | ---- | ------------------- | ------------------ |
  | colors    | r    | Red                 | Rouge              |
  | colors    | y    | Yellow              | Jaune              |
  | colors    | g    | Green               | Vert               |
  | colors    | o    | Other               | Autre              |
  
  ## settings
  
  | form_id   | form_title      | version    | instance_name                               |
  | --------- | --------------- | ---------- | ------------------------------------------- |
  | fav_color | Favorite colors | 2022062100 | concat("Favorite colors - ", ${first_name}) |
  
  '''
# ---
# name: test_excel_to_markdown_snapshot[simple-repeat.xlsx]
  '''
  ## survey
  
  | type         | name        | label                     |
  | ------------ | ----------- | ------------------------- |
  | string       | session_num | Session number            |
  | date         | today       | Today's date              |
  | string       | name        | Your name                 |
  | begin_repeat | observation | Observation               |
  | string       | thing       | What have you observed?   |
  | integer      | count       | How many did you observe? |
  | string       | notes       | Notes                     |
  | end_repeat   | observation |                           |
  
  ## settings
  
  | form_id       | version    |
  | ------------- | ---------- |
  | simple_repeat | 2022101001 |
  
  '''
# ---
# name: test_excel_to_markdown_snapshot[xlsform-dot-org-template.xlsx]
  '''
  ## survey
  
  | type | name | label | hint | required | relevant | appearance | default | constraint | constraint_message | calculation | trigger | choice_filter | parameters | repeat_count | note | image | audio  | video |
  | ---- | ---- | ----- | ---- | -------- | -------- | ---------- | ------- | ---------- | ------------------ | ----------- | ------- | ------------- | ---------- | ------------ | ---- | ----- | ------ | ----- |
  
  ## choices
  
  | list_name | name | label |
  | --------- | ---- | ----- |
  
  ## settings
  
  | form_title | form_id | version                        | instance_name | default_language | style |
  | ---------- | ------- | ------------------------------ | ------------- | ---------------- | ----- |
  |            |         | =TEXT(NOW(), "yyyymmddhhmmss") |               |                  |       |
  
  '''
# ---
# name: test_excel_to_yaml_snapshot[cascading-select.xlsx]
  '''
  survey:
  - type: select_one states
    name: selected_state
    label: Select a state
  - type: select_one counties
    name: selected_county
    label: Select a county
    choice_filter: state=${selected_state}
  - type: select_one cities
    name: selected_city
    label: Select a city
    choice_filter: county=${selected_county}
  choices:
  - list_name: states
    name: texas
    label: Texas
  - list_name: states
    name: washington
    label: Washington
  - list_name: counties
    name: king
    label: King
    state: washington
  - list_name: counties
    name: pierce
    label: Pierce
    state: washington
  - list_name: counties
    name: brewster
    label: Brewster
    state: texas
  - list_name: counties
    name: cameron
    label: Cameron
    state: texas
  - list_name: cities
    name: brownsville
    label: Brownsville
    state: texas
    county: cameron
  - list_name: cities
    name: harlingen
    label: Harlingen
    state: texas
    county: cameron
  - list_name: cities
    name: seattle
    label: Seattle
    state: washington
    county: king
  - list_name: cities
    name: redmond
    label: Redmond
    state: washington
    county: king
  - list_name: cities
    name: tacoma
    label: Tacoma
    state: washington
    county: pierce
  - list_name: cities
    name: puyallup
    label: Puyallup
    state: washington
    county: pierce
  settings:
  - form_title: Cascading select example
    form_id: cascading_select
  yxf:
    headers:
      survey:
      - type
      - name
      - label
      - choice_filter
      choices:
      - list_name
      - name
      - label
      - state
      - county
      settings:
      - form_title
      - form_id
  
  '''
# ---
# name: test_excel_to_yaml_snapshot[favorite-color.xlsx]
  '''
  survey:
  - type: text
    name: first_name
    label::English (en): What is your first name?
    label::French (fr): Quel est votre prénom?
    required: true()
  - type: integer
    name: age
    label::English (en): What is your age, ${first_name}?
    label::French (fr): Quel âge as-tu, ${first_name}?
    required: true()
  - type: select_one colors
    name: favorite_color
    label::English (en): What is your favorite color?
    label::French (fr): Quelle est votre couleur préférée?
  - type: text
    name: favorite_color_other
    label::English (en): Please specify your favorite color
    label::French (fr): Veuillez spécifier votre couleur préférée
    relevant: selected(${favorite_color}, 'o')
  - type: geopoint
    name: location
    label::English (en): Record your location
    label::French (fr): Enregistrez votre position
    required: true()
    relevant: '${age} >= 18 '
    appearance: hide-input
  - type: audit
    name: audit
    parameters: location-priority=balanced location-min-interval=60 location-max-age=120
  choices:
  - list_name: colors
    name: r
    label::English (en): Red
    label::French (fr): Rouge
  - list_name: colors
    name: y
    label::English (en): Yellow
    label::French (fr): Jaune
  - list_name: colors
    name: g
    label::English (en): Green
    label::French (fr): Vert
  - list_name: colors
    name: o
    label::English (en): Other
    label::French (fr): Autre
  settings:
  - form_id: fav_color
    form_title: Favorite colors
    version: 2022062100
    instance_name: concat("Favorite colors - ", ${first_name})
  yxf:
    headers:
      survey:
      - type
      - name
      - label::English (en)
      - label::French (fr)
      - required
      - relevant
      - parameters
      - appearance
      choices:
      - list_name
      - name
      - label::English (en)
      - label::French (fr)
      settings:
      - form_id
      - form_title
      - version
      - instance_name
  
  '''
# ---
# name: test_excel_to_yaml_snapshot[simple-repeat.xlsx]
  '''
  survey:
  - type: string
    name: session_num
    label: Session number
  - type: date
    name: today
    label: Today's date
  - type: string
    name: name
    label: Your name
  - type: begin_repeat
    name: observation
    label: Observation
  - type: string
    name: thing
    label: What have you observed?
  - type: integer
    name: count
    label: How many did you observe?
  - type: string
    name: notes
    label: Notes
  - type: end_repeat
    name: observation
  settings:
  - form_id: simple_repeat
    version: 2022101001
  yxf:
    headers:
      survey:
      - type
      - name
      - label
      settings:
      - form_id
      - version
  
  '''
# ---
